<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению. Сессия № 2
</center>
Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center>Тема 1. Первичный анализ данных с Pandas</center>
## <center>Практическое задание. Анализ данных пассажиров "Титаника"</center>

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/SbRcpW).**

In [5]:
import numpy as np
import pandas as pd

% matplotlib inline

UsageError: Line magic function `%` not found.


**Считаем данные из файла в память в виде объекта Pandas.DataFrame**

In [6]:
data = pd.read_csv("../../data/titanic_train.csv", index_col="PassengerId")

**Данные представлены в виде таблицы. Посмотрим на первые 5 строк:**

In [7]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [8]:
data.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

**Для примера отберем пассажиров, которые сели в Cherbourg (Embarked=C) и заплатили более 200 у.е. за билет (fare > 200).**

Убедитесь, что Вы понимаете, как эта конструкция работает. <br>
Если нет – посмотрите, как вычисляется выражение в квадратных в скобках.

In [9]:
data[(data["Embarked"] == "C") & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                              
119              0      1  PC 17558  247.5208          B58 B60        C  
259              0      0  PC 17755  512.3292              NaN        C  
300              0      1  PC 17558  247.5208          B58 B60        C  
312              2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
378              0      2    113503  211.5000              C82        C

**Можно отсортировать этих людей по убыванию платы за билет.**

In [10]:
data[(data["Embarked"] == "C") & (data["Fare"] > 200)].sort_values(
    by="Fare", ascending=False
).head()

Survived  Pclass                                   Name     Sex  \
PassengerId                                                                    
259                 1       1                       Ward, Miss. Anna  female   
680                 1       1     Cardeza, Mr. Thomas Drake Martinez    male   
738                 1       1                 Lesurer, Mr. Gustave J    male   
312                 1       1             Ryerson, Miss. Emily Borie  female   
743                 1       1  Ryerson, Miss. Susan Parker "Suzette"  female   

              Age  SibSp  Parch    Ticket      Fare            Cabin Embarked  
PassengerId                                                                    
259          35.0      0      0  PC 17755  512.3292              NaN        C  
680          36.0      0      1  PC 17755  512.3292      B51 B53 B55        C  
738          35.0      0      0  PC 17755  512.3292             B101        C  
312          18.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C  
743          21.0      2      2  PC 17608  262.3750  B57 B59 B63 B66        C

**Пример создания признака.**

In [11]:
def age_category(age):
    """
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    """
    if age < 30:
        return 1
    elif age < 55:
        return 2
    else:
        return 3

In [12]:
age_categories = [age_category(age) for age in data.Age]

In [13]:
data["Age_category"] = age_categories

**Другой способ – через `apply`.**

In [14]:
data["Age_category"] = data["Age"].apply(age_category)

**1. Сколько мужчин / женщин находилось на борту?**
- 412 мужчин и 479 женщин
- 314 мужчин и 577 женщин
- 479 мужчин и 412 женщин
- 577 мужчин и 314 женщин

In [27]:
data.groupby('Sex')['Sex'].count()
# 314 женщин и 577 мужчин на борту

Sex
female    314
male      577
Name: Sex, dtype: int64

In [28]:
data[data['Sex'] == 'female'].count(), \
data[data['Sex'] == 'male'].count()

(Survived        314
 Pclass          314
 Name            314
 Sex             314
 Age             261
 SibSp           314
 Parch           314
 Ticket          314
 Fare            314
 Cabin            97
 Embarked        312
 Age_category    314
 dtype: int64,
 Survived        577
 Pclass          577
 Name            577
 Sex             577
 Age             453
 SibSp           577
 Parch           577
 Ticket          577
 Fare            577
 Cabin           107
 Embarked        577
 Age_category    577
 dtype: int64)

**2. Выведите распределение переменной `Pclass` (социально-экономический статус) и это же распределение, только для мужчин / женщин по отдельности. Сколько было мужчин 2-го класса?**
- 104
- 108
- 112
- 125

In [31]:
data.groupby(['Sex', 'Pclass'])['Sex'].count()  # 108 Мужчин 2-ого класса на борту

Sex     Pclass
female  1          94
        2          76
        3         144
male    1         122
        2         108
        3         347
Name: Sex, dtype: int64

In [32]:
data[(data['Sex'] == 'male') & (data['Pclass'] == 2)].count()

Survived        108
Pclass          108
Name            108
Sex             108
Age              99
SibSp           108
Parch           108
Ticket          108
Fare            108
Cabin             6
Embarked        108
Age_category    108
dtype: int64

**3. Каковы медиана и стандартное отклонение платежей (`Fare`)? Округлите до 2 десятичных знаков.**
- Медиана – 14.45, стандартное отклонение – 49.69
- Медиана – 15.1, стандартное отклонение – 12.15
- Медиана – 13.15, стандартное отклонение – 35.3
- Медиана – 17.43, стандартное отклонение – 39.1

In [18]:
pd.set_option('display.precision', 4)
# Медиана 14.45 и стандартное отклонение 46.69
data['Fare'].median(), \
data['Fare'].std()

(14.4542, 49.6934285971809)

**4. Правда ли, что люди моложе 30 лет выживали чаще, чем люди старше 60 лет? Каковы доли выживших в обеих группах?**
- 22.7% среди молодых и 40.6% среди старых
- 40.6% среди молодых и 22.7% среди старых
- 35.3% среди молодых и 27.4% среди старых
- 27.4% среди молодых и 35.3% среди старых

In [33]:
data[(data['Survived'] == 1) & (data['Age'] > 60)].count() / data[data['Age'] > 60].count() * 100, \
    data[(data['Survived'] == 1) & (data['Age'] < 30)].count() / data[data['Age'] < 30].count() * 100
# 40.6 среди молодых и 22.7 среди старых

(Survived        22.7273
 Pclass          22.7273
 Name            22.7273
 Sex             22.7273
 Age             22.7273
 SibSp           22.7273
 Parch           22.7273
 Ticket          22.7273
 Fare            22.7273
 Cabin           27.2727
 Embarked        19.0476
 Age_category    22.7273
 dtype: float64,
 Survived        40.6250
 Pclass          40.6250
 Name            40.6250
 Sex             40.6250
 Age             40.6250
 SibSp           40.6250
 Parch           40.6250
 Ticket          40.6250
 Fare            40.6250
 Cabin           76.9231
 Embarked        40.6250
 Age_category    40.6250
 dtype: float64)

**5. Правда ли, что женщины выживали чаще мужчин? Каковы доли выживших в обеих группах?**
- 30.2% среди мужчин и 46.2% среди женщин
- 35.7% среди мужчин и 74.2% среди женщин
- 21.1% среди мужчин и 46.2% среди женщин
- 18.9% среди мужчин и 74.2% среди женщин

In [20]:
data[(data['Survived'] == 1) & (data['Sex'] == 'male')].count() / data[data['Sex'] == 'male'].count(), \
data[(data['Survived'] == 1) & (data['Sex'] == 'female')].count() / data[data['Sex'] == 'female'].count()
# 18.9 среди мужчин и 74.2 среди женщин

(Survived        0.1889
 Pclass          0.1889
 Name            0.1889
 Sex             0.1889
 Age             0.2053
 SibSp           0.1889
 Parch           0.1889
 Ticket          0.1889
 Fare            0.1889
 Cabin           0.4206
 Embarked        0.1889
 Age_category    0.1889
 dtype: float64,
 Survived        0.7420
 Pclass          0.7420
 Name            0.7420
 Sex             0.7420
 Age             0.7548
 SibSp           0.7420
 Parch           0.7420
 Ticket          0.7420
 Fare            0.7420
 Cabin           0.9381
 Embarked        0.7404
 Age_category    0.7420
 dtype: float64)

**6. Найдите самое популярное имя среди пассажиров Титаника мужского пола?**
- Charles
- Thomas
- William
- John

In [21]:
def parse_first_name_male(name):
    first = name.str.extract(r"Mrs\.\s+[^(]*\((\w+)", expand=False)
    first.loc[first.isna()] = name.str.extract(r"\.\s+(\w+)", expand=False)
    return first
# Функция для парсинга имени из имени фамилии и тд

In [22]:
parse_first_name_male(data.loc[data['Sex']=='male', 'Name'])

PassengerId
1           Owen
5        William
6          James
7        Timothy
8          Gosta
         ...    
884    Frederick
885        Henry
887       Juozas
890         Karl
891      Patrick
Name: Name, Length: 577, dtype: object

In [23]:
parse_first_name_male(data.loc[data['Sex']=='male', 'Name']).value_counts()
# Самое частое мужское имя - William

William    35
John       25
George     14
Charles    13
Thomas     13
           ..
Ignjac      1
Yoto        1
Austen      1
Mitto       1
Juozas      1
Name: Name, Length: 286, dtype: int64

**7. Сравните графически распределение стоимости билетов и возраста у спасенных и у погибших. Средний возраст погибших выше, верно?**
- Да 
- Нет

In [24]:
data[data['Survived'] == 1].mean(), \
data[data['Survived'] == 0].mean()
# ДА - т.к. средний возраст выживших - 28, погибших - 30

C:\Users\Zetis\AppData\Local\Temp\ipykernel_16920\728863080.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data[data['Survived'] == 1].mean(), \
C:\Users\Zetis\AppData\Local\Temp\ipykernel_16920\728863080.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data[data['Survived'] == 0].mean()


(Survived         1.0000
 Pclass           1.9503
 Age             28.3437
 SibSp            0.4737
 Parch            0.4649
 Fare            48.3954
 Age_category     1.7339
 dtype: float64,
 Survived         0.0000
 Pclass           2.5319
 Age             30.6262
 SibSp            0.5537
 Parch            0.3297
 Fare            22.1179
 Age_category     1.8652
 dtype: float64)

**8. Как отличается средний возраст мужчин / женщин в зависимости от класса обслуживания? Выберите верные утверждения:**
- В среднем мужчины 1-го класса старше 40 лет
- В среднем женщины 1-го класса старше 40 лет
- Мужчины всех классов в среднем старше женщин того же класса
- В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса

In [25]:
data[(data['Sex'] == 'male') & (data['Pclass'] == 1)].mean(), \
data[(data['Sex'] == 'female') & (data['Pclass'] == 1)].mean(), \
data[data['Sex'] == 'male'].mean(), \
data[data['Sex'] == 'female'].mean(), \
data[data['Pclass'] == 1].mean(), \
data[data['Pclass'] == 2].mean(), \
data[data['Pclass'] == 3].mean()
# В среднем мужчины 1-го класса старше 40 лет - ДА (41)
# В среднем женщины 1-го класса старше 40 лет - НЕТ (34)
# Мужчины всех классов в среднем старше женщин того же класса - ДА м(30) ж(27)
# В среднем люди в 1 классе старше, чем во 2-ом, а те старше представителей 3-го класса - ДА 1(38) 2(29) 3(25)

C:\Users\Zetis\AppData\Local\Temp\ipykernel_16920\1513645254.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data[(data['Sex'] == 'male') & (data['Pclass'] == 1)].mean(), \
C:\Users\Zetis\AppData\Local\Temp\ipykernel_16920\1513645254.py:2: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  data[(data['Sex'] == 'female') & (data['Pclass'] == 1)].mean(), \
C:\Users\Zetis\AppData\Local\Temp\ipykernel_16920\1513645254.py:3: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will de

(Survived         0.3689
 Pclass           1.0000
 Age             41.2814
 SibSp            0.3115
 Parch            0.2787
 Fare            67.2261
 Age_category     2.1311
 dtype: float64,
 Survived          0.9681
 Pclass            1.0000
 Age              34.6118
 SibSp             0.5532
 Parch             0.4574
 Fare            106.1258
 Age_category      1.8511
 dtype: float64,
 Survived         0.1889
 Pclass           2.3899
 Age             30.7266
 SibSp            0.4298
 Parch            0.2357
 Fare            25.5239
 Age_category     1.8596
 dtype: float64,
 Survived         0.7420
 Pclass           2.1592
 Age             27.9157
 SibSp            0.6943
 Parch            0.6497
 Fare            44.4798
 Age_category     1.7325
 dtype: float64,
 Survived         0.6296
 Pclass           1.0000
 Age             38.2334
 SibSp            0.4167
 Parch            0.3565
 Fare            84.1547
 Age_category     2.0093
 dtype: float64,
 Survived         0.4728
 Pclass 